*: update document of syncer (#899)
pingcap/docs: tools/syncer.md

Description of Syncer command line options:

```
Usage of syncer:
  -L string
        log level: debug, info, warn, error, fatal (default "info")

>   -V    to print Syncer version info (default false)

Reviewer: not a new line?

  -b int
        the size of batch transactions (default 10)
  -c int
        the number of batch threads that Syncer processes (default 16)
  -config string
        path to config file
  -enable-ansi-quotes

>         to specify the corresponding configuration file when starting Syncer; for example, `--config config.toml`

Reviewer: enable ansi-quotes sql_mode?

  -enable-gtid
        to start Syncer using the mode; default false; before enabling this option, you need to enable GTID in the upstream MySQL
  -flavor string

>         use flavor for different MySQL source versions; support "mysql", "mariadb" now; if you replicate from mariadb, please set it to "mariadb" (default "mysql")

Reviewer 2: ```suggestion
        use flavor for different MySQL source versions; support "mysql", "mariadb" now; if you replicate data from MariaDB, set it to "mariadb" (default "mysql")
```

  -log-file string
        to specify the log file directory, such as `--log-file ./syncer.log`
  -log-rotate string
        to specify the log file rotating cycle, hour/day (default "day")
  -max-retry int

>         to specify maxinum retry when network interruption (default 100)

Reviewer 2: ```suggestion
        to specify maximum retry times during network interruption (default 100)
```

  -meta string

>         to specify the meta file of Syncer upstream (in the same directory with the configuration file by default "syncer.meta")

Reviewer 2: ```suggestion
        to specify the meta file of the upstream of Syncer (in the same directory with the configuration file, "syncer.meta" by default)
```

  -persistent-dir string

>         to specify  syncer history table structures persistent dir; set to non-empty string will choosing history table structure according to column length when constructing DML

Reviewer: in fact, it's *file* path, not *dir*.

Author: Users are bound to keep wondering and asking

>         to specify syncer history table structures persistent file (historical reason: it's not directory); set to non-empty string will choosing history table structure according to column length when constructing DML

Reviewer 2: ```suggestion
        to specify the persistent file (historical reason: it is not a directory) of Syncer history table schemas; if you set it to a non-empty string, the history table schema is chosen according to the column length when you construct DML statements
```

  -safe-mode

>         to specify enable safe mode to make syncer reentrant

Reviewer 2: ```suggestion
        to specify and enable the safe mode to make Syncer reentrant
```

  -server-id int
        to specify MySQL slave sever-id (default 101)
  -status-addr string
        to specify Syncer metrics (default :8271), such as `--status-addr 127:0.0.1:8271`
```

The `config.toml` configuration file of Syncer:

```toml
log-level = "info"
log-file = "syncer.log"
log-rotate = "day"

server-id = 101

# The file path for meta:
meta = "./syncer.meta"
worker-count = 16
batch = 1000
flavor = "mysql"

# The testing address for pprof. It can also be used by Prometheus to pull Syncer metrics.
status-addr = ":8271"

> # If set true, syncer will stop and exit on ddl.

Reviewer 2: ```suggestion
# If set it to true, Syncer will stop and exit when it encounters the DDL operation.
```

stop-on-ddl = false

> # max-retry is used for retry when network interruption.

Reviewer 2: ```suggestion
# max-retry is used for retry during network interruption.
```

> # If set it to true, Syncer will stop and exit when it encounters the DDL operation.

Reviewer 2: ```suggestion
# If you set its value to true, Syncer stops and exits when it encounters the DDL operation.
```

> #### title: binlog events

Author: @lilin90 @csuzhangxc any problem with these metric descriptions, if you don't have any problems, I would also fill them in syncer monitor template

Reviewer: no problem.

> - metrics: `rate(syncer_binlog_event_count[1m])`

Reviewer: seems <https://github.com/pingcap/docs/blob/master/etc/Syncer.json> still not updated?

- info: QPS of the binlog event that has been recieved by Syncer

#### title: binlog event transform

- metrics: `histogram_quantile(0.8, sum(rate(syncer_binlog_event_bucket[1m])) by (le))`

> - info: the cost of transforming binlog event to SQLs by Syncer

Reviewer 2: ```suggestion
- info: the cost of transforming binlog event to SQL statements by Syncer
```

#### title: transaction latency

- metrics: `histogram_quantile(0.95, sum(rate(syncer_txn_cost_in_second_bucket[1m])) by (le))`
- info: the cost of executing a transaction on TiDB

#### title: transaction tps

- metrics: `rate(syncer_txn_cost_in_second_count[1m])`
- info: TPS of executing a transaction on TiDB

#### title: binlog file gap

- metrics: `syncer_binlog_file{node="master"} - ON(instance, job) syncer_binlog_file{node="syncer"}`
- info: the number of different binlog files between the upstream and the downstream in the process of synchronization; the normal value is 0, which indicates real-time synchronization; a larger value indicates a larger number of binlog files discrepancy

#### title: binlog skipped events

- metrics: `rate(syncer_binlog_skipped_events_total[1m])`
- info: the total number of SQL statements that Syncer skips when the upstream synchronizes binlog files with the downstream; you can configure the format of SQL statements skipped by Syncer using the `skip-sqls` parameter in the `syncer.toml` file.

#### title: execution jobs

- metrics: `sum(rate(syncer_add_jobs_total[1m])) by (queueNo)`

> - info: the count of job that had been added into execution queue

Reviewer 2: ```suggestion
- info: the count of jobs that have been added into the execution queue
```

> #### title: pending  jobs

Reviewer 2: ```suggestion
#### title: pending jobs
```

- metrics: `sum(rate(syncer_add_jobs_total[1m]) - rate(syncer_finished_jobs_total[1m])) by (queueNo)`

> - info: the count of job that had been applied into TiDB

Reviewer 2: ```suggestion
- info: the count of jobs that have been applied into TiDB
```

#### title: binlog events

- metrics: `rate(syncer_binlog_event_count[1m])`

> - info: QPS of the binlog event that has been recieved by Syncer

Reviewer 2: ```suggestion
- info: QPS of the binlog event that has been received by Syncer
```

#### title: binlog event transform

- metrics: `histogram_quantile(0.8, sum(rate(syncer_binlog_event_bucket[1m])) by (le))`

> - info: the cost of transforming binlog event to SQL statements by Syncer

Reviewer 2: ```suggestion
- info: the cost of transforming the binlog event to SQL statements by Syncer
```

#### title: transaction latency

- metrics: `histogram_quantile(0.95, sum(rate(syncer_txn_cost_in_second_bucket[1m])) by (le))`
- info: the cost of executing a transaction on TiDB

#### title: transaction tps

- metrics: `rate(syncer_txn_cost_in_second_count[1m])`
- info: TPS of executing a transaction on TiDB

#### title: binlog file gap

- metrics: `syncer_binlog_file{node="master"} - ON(instance, job) syncer_binlog_file{node="syncer"}`
- info: the number of different binlog files between the upstream and the downstream in the process of replication; the normal value is 0, which indicates real-time replication; a larger value indicates a larger number of binlog files discrepancy

#### title: binlog skipped events

- metrics: `rate(syncer_binlog_skipped_events_total[1m])`

> - info: the total number of SQL statements that Syncer skips when the upstream synchronizes binlog files with the downstream; you can configure the format of SQL statements skipped by Syncer using the `skip-sqls` parameter in the `syncer.toml` file.

Reviewer 3: `skip-sqls` is deprecated, should we change this to `skip-ddls` and `skip-dmls`

#### title: position binlog position

- metrics: `syncer_binlog_pos{node="syncer"}` and `syncer_binlog_pos{node="master"}`

> - info: it works with `file number of binlog position`. `syncer_binlog_pos{node="master"}` indicate the position of latest binlog position that fetch from MySQL, and `syncer_binlog_pos{node="syncer"}` indicate the position of binlog position syncer had executed

Reviewer 2: ```suggestion
- info: it works with `file number of binlog position`. `syncer_binlog_pos{node="master"}` indicates the position of latest binlog position fetched from MySQL, and `syncer_binlog_pos{node="syncer"}` indicates the position of the binlog position that Syncer has replicated.
```

#### title: file number of binlog position

- metrics: `syncer_binlog_file{node="syncer"}` and `syncer_binlog_file{node="master"}`

> - info: it works with `position of binlog position`. `syncer_binlog_file{node="master"}` indicate the file number of latest binlog position that fetch from MySQL, and `syncer_binlog_file{node="syncer"}` indicate the file number of binlog position syncer had executed

Reviewer 2: ```suggestion
- info: it works with `position of binlog position`. `syncer_binlog_file{node="master"}` indicates the file number of the latest binlog position fetched from MySQL, and `syncer_binlog_file{node="syncer"}` indicates the file number of the binlog position that Syncer has replicated.
```
